sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py

```diff
@@ -167,6 +174,147 @@ def compute(
         """
 
 
+class PairwiseDistances(BaseDistancesReductionDispatcher):
+    """Compute the pairwise distances matrix for two sets of vectors.
+
+    The distance function `dist` depends on the values of the `metric`
+    and `metric_kwargs` parameters.
+
+    This class only computes the pairwise distances matrix without
+    applying any reduction on it. It shares most of the underlying
+    code infrastructure with reducing variants to leverage multi-thread
+    parallelism. However contrary to the reducing variants, no chunking
+    is applied to allow for contiguous write access to the final distance
+    array that is not expected to fit in the CPU cache in general.
+
+    This class is not meant to be instantiated, one should only use
+    its :meth:`compute` classmethod which handles allocation and
+    deallocation consistently.
+    """
+
+    @classmethod
+    def is_usable_for(cls, X, Y, metric, metric_kwargs=None) -> bool:
+
+        effective_n_threads = _openmp_effective_n_threads()
+
+        def is_euclidean(metric, metric_kwargs):
+            metric_kwargs = metric_kwargs or dict()
+            euclidean_metrics = [
+                "euclidean",
+                "sqeuclidean",
+                "l2",
+            ]
+            return metric in euclidean_metrics or (
+                metric == "minkowski" and metric_kwargs.get("p") == 2
```

Comment:
I'm not 100% confident about getting `p` through `metric_kwargs` since we usually pass it as a standalone parameter (in the `NearestNeighbors` example). WDYT?

Reply:
To make sure `p` is correctly used in `PairwiseDistance` I suggest this test, WDYT?
```python
kwds = {"p": 2.0}
S = pairwise_distances(X, metric="minkowski", **kwds)

kwds = {"p": 3.0}
S3 = pairwise_distances(X, metric="minkowski", **kwds)
with pytest.raises(ValueError):
    assert_allclose(S, S3)
```

When `p=3`, `PairwiseDistance.is_usable_for()` returns `True`, and this value of `p` is used in the downstream `DistanceMetric`.